devito/tools/data_structures.py

```diff
@@ -467,6 +467,29 @@ def topological_sort(self, choose_element=None):
         else:
             raise ValueError('graph is not acyclic')
 
+    def connected_components(self, enumerated=False):
+        """
+        Find all connected sub-graphs and return them as a list.
+        """
+        groups = []
+
+        for n0 in self.graph:
+            found = {n0} | set(self.all_downstreams(n0))
+            for g in groups:
+                if g.intersection(found):
+                    g.update(found)
+                    break
+            else:
```

Comment:
what if does this else go with?

Reply:
It's actually part of the `for` loop construct. I've learnt about Python's else for `for` loops only recently, and I've started using it anywhere I can because it's really extremely handy to make the code easier to read.

See here: https://book.pythontips.com/en/latest/for_-_else.html#else-clause

Reply:
Interesting, didn't know those existed that's quite handy